lib/TPP/Transforms/LoopExpansion.cpp

```diff
@@ -0,0 +1,143 @@
+//===- LoopExpansion.cpp -----------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file  splits parallel loop into scf fors.
+//
+//===----------------------------------------------------------------------===//
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DECL_LOOPEXPANSIONPASS
+#define GEN_PASS_DEF_LOOPEXPANSIONPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::scf;
+using namespace std;
+
+namespace mlir {
+namespace tpp {
+
+struct LoopExpansion : OpRewritePattern<scf::ForallOp> {
```

Comment:
Just curious, why `scf.forall` and not `scf.parallel`?

No preference either way for me.

Reply:
Forall is updated in the loopinsertion pass, hence we expand it in this pass. There's also a transform pass that expands forall to for that we eventually want to use instead of this pass. That's the reason we're using forall instead of parallel.